File: typescript/src/types/binary-serializer-char.ts
import { DataReader } from '../buffers/data-reader.interface';
import { DataWriter } from '../buffers/data-writer.interface';
import { BinaryTypeSerializer } from './binary-type-serializer.interface';

/**
 * Reads (and writes) characters to (and from) a binary data source.
 */
export class BinarySerializerChar implements BinaryTypeSerializer<string> {
    get sizeInBytes(): number {
        return 2;
    }

    encode(writer: DataWriter, value: string): void {
        if (value.length !== 1) {
            throw new Error('Value must be a single character');
        }

        const charCode = value.charCodeAt(0);
        const bytes = new Uint8Array(this.sizeInBytes);
        new DataView(bytes.buffer).setUint16(0, charCode, true);
        
        writer.writeBytes(bytes);
    }

    decode(reader: DataReader): string {
        const bytes = reader.readBytes(this.sizeInBytes);
        const charCode = new DataView(bytes.buffer).getUint16(0, true);
        
        return String.fromCharCode(charCode);
    }

    getEquals(a: string, b: string): boolean {
        if (a.length !== 1 || b.length !== 1) {
            throw new Error('Values must be single characters.');
        }
        return a === b;
    }
}

File: dotnet/Barchart.BinarySerializer/Types/BinarySerializerChar.cs
#region Using Statements

using Barchart.BinarySerializer.Buffers;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) characters to (and from) a binary data source.
/// </summary>
public class BinarySerializerChar : IBinaryTypeSerializer<char>
{
    #region Constants

    private const int ENCODED_LENGTH_IN_BYTES = sizeof(char);
    
    #endregion

    #region Methods
    
    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, char value) 
    {
        writer.WriteBytes(BitConverter.GetBytes(value));
    }

    /// <inheritdoc />
    public char Decode(IDataBufferReader reader)
    {
        return BitConverter.ToChar(reader.ReadBytes(ENCODED_LENGTH_IN_BYTES));
    }
    
    /// <inheritdoc />
    public bool GetEquals(char a, char b)
    {
        return a.Equals(b);
    }

    #endregion
}